### PR TITLE
[Dataset] Add WikiCS Dataset

### PR DESCRIPTION
**Pull request by RecLusIve-F**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/data/wikics.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+import numpy as np
+from .. import backend as F
+from ..convert import graph
+from .dgl_dataset import DGLBuiltinDataset
+from ..transforms import to_bidirected, reorder_graph
+from .utils import generate_mask_tensor, load_graphs, save_graphs, _get_dgl_url
+
+
+class WikiCSDataset(DGLBuiltinDataset):
+    r"""Wiki-CS is a Wikipedia-based dataset for node classification.
```
Comment (mufeili) [inline python/dgl/data/wikics.py:14]: Add "from `Wiki-CS: A Wikipedia-Based Benchmark for Graph Neural Networks <https://arxiv.org/abs/2109.01116>`_"

File: python/dgl/data/wikics.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+    The dataset consists of nodes corresponding to Computer Science articles, with edges based on
+    hyperlinks and 10 classes representing different branches of the field.
+
+    Reference: `<https://arxiv.org/abs/2109.01116>`_
```
Comment (mufeili) [inline python/dgl/data/wikics.py:19]: Remove this line.

File: python/dgl/data/wikics.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+    WikiCS dataset statistics:
+
+    - Nodes: 11,701
+    - Edges: 431,726 (note that the original dataset has 216,123 edges but DGL adds
+      the reverse edges and remove the duplicates, hence with a different number)
```
Comment (mufeili) [inline python/dgl/data/wikics.py:25]: "remove the duplicates" -> "removes the duplicate edges"

File: python/dgl/data/wikics.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    - Number of Classes: 10
+    - Node feature size:: 300
```
Comment (mufeili) [inline python/dgl/data/wikics.py:27]: extra ":"?

File: python/dgl/data/wikics.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    - Number of different train, validation, stopping splits: 20
```
Comment (mufeili) [inline python/dgl/data/wikics.py:28]: It seems that you also have a test subset. What does "stopping" stand for here?

File: python/dgl/data/wikics.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    ----------
+    num_classes : int
+        Number of node classes
+
+    Examples
+    --------
+    >>> dataset = WikiCSDataset()
+    >>> len(dataset)
+    1
+    >>> for g in dataset:
```
Comment (mufeili) [inline python/dgl/data/wikics.py:57]: Since the dataset has a single graph, use `g = dataset[0]` instead.

File: python/dgl/data/wikics.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    ....    # get edge feature
+    ....    feat = g.ndata['feat']
+    ....    # your code here
```
Comment (mufeili) [inline python/dgl/data/wikics.py:60]: How to retrieve the node labels?

File: python/dgl/data/wikics.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+        super(WikiCSDataset, self).__init__(name='wiki_cs',
+                                            raw_dir=raw_dir,
+                                            url=_url,
+                                            force_reload=force_reload,
+                                            verbose=verbose,
+                                            transform=transform)
+
+    def process(self):
+        """process raw data to graph, labels and masks"""
+        data = json.load(open(os.path.join(self.raw_path, 'data.json')))
```
Comment (mufeili) [inline python/dgl/data/wikics.py:74]: Use

```
with open(...):
    ...
```

File: python/dgl/data/wikics.py [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+        features = F.tensor(np.array(data['features']), dtype=F.data_type_dict['float32'])
```
Comment (mufeili) [inline python/dgl/data/wikics.py:75]: You can simply use `F.float32`.

File: python/dgl/data/wikics.py [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+        labels = F.tensor(np.array(data['labels']), dtype=F.data_type_dict['int64'])
```
Comment (mufeili) [inline python/dgl/data/wikics.py:76]: Ditto

File: python/dgl/data/wikics.py [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+        :class:`dgl.DGLGraph`
+
+            graph structure, labels, features and masks.
+
+            - ``ndata['label']``: ground truth labels
+            - ``ndata['feat']``: node features
+            - ``ndata['train_mask']``: training mask
+            - ``ndata['val_mask']``: validation mask
+            - ``ndata['stopping_mask']``: stopping mask
+            - ``ndata['test_mask']``: test mask
```
Comment (mufeili) [inline python/dgl/data/wikics.py:141]: Perhaps you can replace L136-141 with a few lines of code in the example above.

**Comment by mufeili** [comment]
You also need to resolve the conflicts reported by github.

File: python/dgl/data/wikics.py (hunk 5 above)
Comment (RecLusIve-F) [reply python/dgl/data/wikics.py:28]: In the paper, the authors suggest using the validation set for hyperparameter tuning, the stop set for early stopping criterion, and the test set for model evaluation.

Comment (mufeili) [reply python/dgl/data/wikics.py:28]: Is this introduced in "https://arxiv.org/abs/2109.01116"? Does the training set, validation set, test set, and stop set form a partition? Or is one set a subset of another set?

Comment (RecLusIve-F) [reply python/dgl/data/wikics.py:28]: This is introduced in "https://arxiv.org/abs/2007.02901v2". They split the nodes in each class into two sets, 50% for the test set and 50% potentially visible. From the visible set, they generated 20 different splits of training, validation and early-stopping sets: 5% of the nodes in each class were used for training in each split, 22.5% were used to evaluate the early-stopping criterion, and 22.5% were used as the validation set for hyperparameter tuning.

Comment (mufeili) [reply python/dgl/data/wikics.py:28]: What's the shape of `g.ndata['train_mask']`? Are these random splits now deterministic?

Comment (mufeili) [reply python/dgl/data/wikics.py:27]: I see. Then perhaps you need to add a description of the shape of these masks and what they stand for.

Comment (RecLusIve-F) [reply python/dgl/data/wikics.py:27]: Got it

File: python/dgl/data/wikics.py [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+    >>> feat = g.ndata['feat']
+    >>> # get node labels
+    >>> labels = g.ndata['label']
+    >>> # get data split
+    >>> train_mask = g.ndata['train_masks']
+    >>> val_mask = g.ndata['val_masks']
+    >>> stopping_mask = g.ndata['stopping_masks']
+    >>> test_mask = g.ndata['test_mask']
+    >>>
+    >>> # Train, Validation and Test
```
Comment (mufeili) [inline python/dgl/data/wikics.py:67]: L66-67 can be removed.

File: python/dgl/data/wikics.py [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+    >>> g = dataset[0]
+    >>> # get node feature
+    >>> feat = g.ndata['feat']
+    >>> # get node labels
+    >>> labels = g.ndata['label']
+    >>> # get data split
+    >>> train_mask = g.ndata['train_masks']
+    >>> val_mask = g.ndata['val_masks']
+    >>> stopping_mask = g.ndata['stopping_masks']
+    >>> test_mask = g.ndata['test_mask']
```
Comment (mufeili) [inline python/dgl/data/wikics.py:65]: Describe the shape of the mask tensors. For tensors of shape (*, 20), explain what `20` stands for.

**Review by mufeili — APPROVED** [review]
I'm good other than some final minor comments.
